src/v/storage/log_reader.cc

```diff
@@ -27,11 +27,9 @@ batch_consumer::consume_result skipping_consumer::accept_batch_start(
   const model::record_batch_header& header) const {
     // check for holes in the offset range on disk
     // skip check for compacted logs
-    if (unlikely(
-          !_reader._seg.is_compacted_segment() && _expected_next_batch() >= 0
-          && header.base_offset != _expected_next_batch)) {
+    if (unlikely(header.base_offset < _expected_next_batch)) {
         throw std::runtime_error(fmt::format(
-          "hole encountered reading from disk log: "
+          "incorrect offset encountered reading from disk log: "
```

Comment:
looks ok. ideally we would track per segment state about if a segment has been compacted or not and apply the rule based on that. since compacted or not-compacted is a property of a segment independent of the current topic policy it would work even if we switched from compaction to deletion.